server/log/main.go

```diff
@@ -32,12 +33,39 @@ var batchSizeFlag = flag.Int("batch_size", 50, "Max number of leaves to process
 var privateKeyFile = flag.String("private_key_file", "", "File containing a PEM encoded private key")
 var privateKeyPassword = flag.String("private_key_password", "", "Password for server private key")
 
-// TODO(Martin2112): Needs a more realistic provider of log storage with some caching
-// and ability to swap out for different storage type
+// Must hold this lock before accessing the storage map
+var storageMapGuard sync.Mutex
+// Map from tree ID to storage impl for that log
+var storageMap = make(map[int64]storage.LogStorage)
+
+// TODO(Martin2112): Needs to be able to swap out for different storage type
 func simpleMySqlStorageProvider(treeID int64) (storage.LogStorage, error) {
 	return mysql.NewLogStorage(trillian.LogID{[]byte("TODO"), treeID}, *mysqlUriFlag)
 }
 
+// TODO(Martin2112): Could pull this out as a wrapper so it can be used elsewhere
+func getStorageForLog(logId int64) (storage.LogStorage, error) {
+	storageMapGuard.Lock()
+	defer storageMapGuard.Unlock()
+
+	s, ok := storageMap[logId]
+
+	if !ok {
+		glog.Infof("Creating new storage for log: %d", logId)
+
+		s, err := simpleMySqlStorageProvider(logId)
+
+		if err != nil {
+			return s, err
+		}
+
+		storageMap[logId] = s
+		return s, nil
```

Comment:
Can ding this return and fall-through, though you might need to have an explicit
```go
var err error
```
and change L51 to use `=` rather than `:=`